### PR TITLE
Exception handling in PipeUtil::read()

### PR DESCRIPTION
**Pull request by navinrathore**
Here is a replaced Fraft PR to handle PipeUtil::Read() exception handling.

* The ZinggClientException has been used for the purpose. Should there be different exception used? If yes, it should better be derived from Exception
* The two variations. read() and readWithException() may not be needed. As our code is already handling these scenario using catching "Exception" and do nothing. single read() with exception is fine.
* Use of ZinggClientException in function signatures is done to satisfy exception throw mechanism and compiler
* error messages like below is printed on cli.
``` 
zingg.client.Client - Apologies for this message. Zingg has encountered an error. Path does not exist: file:/home/work/product/final/zingg-1/models/900/trainingData/marked
```
Please provide your inputs on the approach.
Note: Though major flows have been tested. Still more testing required.

### CODE REVIEW

File: core/src/main/java/zingg/util/PipeUtil.java [hunk 1]
```diff
@@ -68,22 +69,26 @@ private static DataFrameReader getReader(SparkSession spark, Pipe p) {
...
 		Dataset<Row> input = null;
 		LOG.warn("Reading " + p);
-		if (p.getProps().containsKey(FilePipe.LOCATION)) {
-			input = reader.load(p.get(FilePipe.LOCATION));
-		}
-		else {
-			input = reader.load();
-		}
-		if (addSource) {
-			input = input.withColumn(ColName.SOURCE_COL, functions.lit(p.getName()));
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/util/PipeUtil.java:81]: We aren't throwing any new exception here. Catching one and throwing another.
So if an exception is already thrown (that is indeed done in spark.read() in this case), it is handled somewhere (first catch block). So with this change, there will not be any impact in program behavior. Just things made explicit to have the exception handled by the caller. the specific type of the exception. Else, we are loosing the error message in the process.

Still, we have to see which one to use 'Exception' or 'ZinggClientException' in the flow.
We are already handling Exception in top level classes. Why are we loosing ex.getMessage(). I note that intermediate function calls do not h ave explicit exception specification.

[the remark this answers is not in the document]
Comment (sonalgoyal) [reply core/src/main/java/zingg/util/PipeUtil.java:81]: ok, makes sense. but the message here has to be that we could not read the <pipe info> and also wrap the actual exception thrown

File: core/src/main/java/zingg/Labeller.java [hunk 2]
```diff
@@ -75,11 +75,11 @@ protected void getMarkedRecordsStat(Dataset<Row> markedRecords) {
 
 	public void processRecordsCli(Dataset<Row> lines) throws ZinggClientException {
 		LOG.info("Processing Records for CLI Labelling");
-		printMarkedRecordsStat();
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/Labeller.java:78]: removed **return statement.**

File: core/src/main/java/zingg/util/PipeUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/util/PipeUtil.java:81]: Used wrapped exception.

File: core/src/main/java/zingg/Matcher.java [hunk 3]
```diff
@@ -41,11 +41,12 @@ public Matcher() {
         setZinggOptions(ZinggOptions.MATCH);
     }
 
-	protected Dataset<Row> getTestData() {
-		return PipeUtil.read(spark, true, args.getNumPartitions(), true, args.getData());
+	protected Dataset<Row> getTestData() throws ZinggClientException{
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/Matcher.java:44]: To satisfy the requirement of catching or throwing the **ZinggClientException**. All the intermediate functions are mandating to add this.
In this regard, are types **Exception and ZinggClientException** different?

File: core/src/main/java/zingg/Labeller.java [hunk 4]
```diff
@@ -51,7 +51,7 @@ public Dataset<Row> getUnmarkedRecords() throws ZinggClientException {
 			unmarkedRecords = PipeUtil.read(spark, false, false, PipeUtil.getTrainingDataUnmarkedPipe(args));
 			try {
 				markedRecords = PipeUtil.read(spark, false, false, PipeUtil.getTrainingDataMarkedPipe(args));
-			} catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/Labeller.java:54]: read() throws only **ZinggClientException.**
Earlier, why didnot read() mandate to ctch **Exception** everywhere it is called. Is there any difference in **Exception and ZinggClientException**?

File: core/src/main/java/zingg/LabelUpdater.java [hunk 5]
```diff
@@ -41,12 +41,12 @@ public void execute() throws ZinggClientException {
 
 	public void processRecordsCli(Dataset<Row> lines) throws ZinggClientException {
 		LOG.info("Processing Records for CLI updateLabelling");
-		getMarkedRecordsStat(lines);
-		printMarkedRecordsStat();
 		if (lines == null || lines.count() == 0) {
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/LabelUpdater.java]: Removed **Return**.
Earlier thought that following code block was already very big. It should not be cluttered further.

File: core/src/main/java/zingg/Labeller.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sonalgoyal) [reply core/src/main/java/zingg/Labeller.java:54]: yes ZCE extends from throwable

File: core/src/main/java/zingg/LabelUpdater.java [hunk 6]
```diff
@@ -41,70 +41,68 @@ public void execute() throws ZinggClientException {
 
 	public void processRecordsCli(Dataset<Row> lines) throws ZinggClientException {
 		LOG.info("Processing Records for CLI updateLabelling");
-		getMarkedRecordsStat(lines);
-		printMarkedRecordsStat();
-		if (lines == null || lines.count() == 0) {
-			LOG.info("There is no marked record for updating. Please run findTrainingData/label jobs to generate training data.");
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply core/src/main/java/zingg/LabelUpdater.java:47]: Yes. made changes in Labeller and UpdateLabeller.

File: core/src/main/java/zingg/util/PipeUtil.java [hunk 7]
```diff
@@ -68,22 +69,27 @@ private static DataFrameReader getReader(SparkSession spark, Pipe p) {
...
+			else {
+				input = reader.load();
+			}
+			if (addSource) {
+				input = input.withColumn(ColName.SOURCE_COL, functions.lit(p.getName()));
+			}
+		} catch (Exception ex) {
+			LOG.warn(ex.getMessage());
+			throw new ZinggClientException("Could not read data.", ex);
 		}
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/util/PipeUtil.java:88]: does this print which pipe it could not read?

Comment (navinrathore) [reply core/src/main/java/zingg/util/PipeUtil.java:88]: Yes. I think at line number 74, pipe information is logged in all cases.

**Pull request merged**